### PR TITLE
Observable collection

### PR DESCRIPTION
**Pull request by SharonGoliath**
For Metrics and Rejected items.

### CODE REVIEW

File: caom2utils/caom2utils/tests/test_wcsvalidator.py [hunk 1]
```diff
@@ -96,22 +96,22 @@ def test_bad_temporalwcs(self):
         bad_temporal_wcs = TimeTestUtil.bad_ctype_wcs()
         with pytest.raises(InvalidWCSError) as ex:
             wcsvalidator._validate_temporal_wcs(bad_temporal_wcs)
-        assert('unexpected TIMESYS, CTYPE' in str(ex))
+        assert 'unexpected TIMESYS, CTYPE' in repr(ex.value)
```
Comment (andamian) [inline caom2utils/caom2utils/tests/test_wcsvalidator.py]: `str(ex)` not working? That worries me as I've used a lot before...

**Comment by andamian** [comment]
Please update the documentation of package either in the 
README of the package which is displayed on `github`
or in the `__init__.py` which is returned when users do `help(caom2pipe)`. 
An overall view of the architecture would also be helpful.

File: caom2pipe/caom2pipe/execute_composable.py [hunk 2]
```diff
@@ -225,6 +239,25 @@ def fname_on_disk(self):
...
+
+    @property
+    def lineage(self):
+        """The value provided to the --lineage parameter for
+        fits2caom2 extensions."""
+        return '{}/{}'.format(self.product_id, self.file_uri)
+
+    @property
+    def external_urls(self):
+        return None
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/execute_composable.py]: Yes, as only gemini retrieves metadata from external URLs.

File: caom2pipe/caom2pipe/execute_composable.py [hunk 3]
```diff
@@ -317,9 +350,10 @@ def __init__(self, config, task_type, storage_name, command_name,
             handler.setLevel(config.logging_level)
             handler.setFormatter(formatter)
         self.logging_level_param = self._set_logging_level_param(
-            config.logging_level)
+            config.logging_level)[0]
+        self.log_level_as = self._set_logging_level_param(
+            config.logging_level)[1]
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/execute_composable.py]: Renamed to '_specify_logging_level_param', and consumed on one line.

File: caom2pipe/caom2pipe/execute_composable.py [hunk 4]
```diff
@@ -526,42 +529,46 @@ def _visit_meta(self, observation):
...
+                      'observable': self.observable}
             for visitor in self.meta_visitors:
                 try:
                     self.logger.debug('Visit for {}'.format(visitor))
                     visitor.visit(observation, **kwargs)
                 except Exception as e:
                     raise mc.CadcException(e)
 
+    @staticmethod
+    def _set_external_urls_param(external_urls):
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/execute_composable.py]: Renamed to '_specify_external_urls_param'.

File: caom2pipe/caom2pipe/execute_composable.py [hunk 5]
```diff
@@ -526,42 +529,46 @@ def _visit_meta(self, observation):
...
+        """Make a list of external urls into a command-line parameter."""
+        if external_urls is None:
+            result = ''
+        else:
+            result = '--external_url {}'.format(external_urls)
+        return result
+
     @staticmethod
     def _set_logging_level_param(logging_level):
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/execute_composable.py]: Done.

File: caom2pipe/caom2pipe/manage_composable.py [hunk 6]
```diff
@@ -72,32 +72,39 @@
...
-           'compare_checksum_client', 'Features', 'write_to_file',
+           'Features', 'write_to_file',
            'read_from_file', 'read_file_list_from_archive', 'update_typed_set',
            'get_cadc_headers', 'get_lineage', 'get_artifact_metadata',
-           'data_put', 'data_get', 'build_uri']
+           'data_put', 'data_get', 'build_uri', 'make_seconds',
+           'increment_time', 'ISO_8601_FORMAT', 'http_get', 'Rejected',
+           'record_progress', 'Work', 'look_pull_and_put', 'Observable',
+           'Metrics', 'repo_create', 'repo_delete', 'repo_get',
+           'repo_update']
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/manage_composable.py]: In general, the functionality in manage_composable, caom_composable, and astro_composable should be available for any collection pipeline implementation. That does it make it a wide API however. Is there a better way to organize this?

File: caom2pipe/caom2pipe/manage_composable.py [hunk 7]
```diff
@@ -226,52 +244,227 @@ def save_state(self, key, value):
...
         else:
             self.logger.warning('No bookmarks found for {}'.format(key))
 
 
+class Work(object):
+    """"Abstract-like class that defines the operations used to chunk work when
+    controlling execution by State."""
+
+    def __init__(self, max_ts_s):
+        self._max_ts_s = max_ts_s
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/manage_composable.py]: Description added to property.

File: caom2pipe/caom2pipe/manage_composable.py [hunk 8]
```diff
@@ -226,52 +244,227 @@ def save_state(self, key, value):
...
+
+    @property
+    def max_ts_s(self):
+        return self._max_ts_s
+
+    def initialize(self):
+        """Anything necessary to make todo work."""
+        pass
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/manage_composable.py]: Done.

File: caom2pipe/caom2pipe/manage_composable.py [hunk 9]
```diff
@@ -226,52 +244,227 @@ def save_state(self, key, value):
...
+            self.content = {}
+        for reason in Rejected.reasons.keys():
+            if reason not in self.content:
+                self.content[reason] = []
+
+    def check_and_record(self, message, obs_id):
+        """Keep track of an additional entry."""
+        for reason, reason_str in Rejected.reasons.items():
+            if reason_str in message:
+                self.content[reason].append(obs_id)
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/manage_composable.py]: Done.

File: caom2pipe/caom2pipe/manage_composable.py [hunk 10]
```diff
@@ -226,52 +244,227 @@ def save_state(self, key, value):
...
+
+
+class Observable(object):
+    """
+    A class to contain all the classes that maintain information between
+    pipeline execution instances.
+    """
+
+    def __init__(self, rejected, metrics):
+        self._rejected = rejected
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/manage_composable.py]: Added.

File: caom2pipe/caom2pipe/manage_composable.py [hunk 11]
```diff
@@ -226,52 +244,227 @@ def save_state(self, key, value):
...
+#         if enabled:
+#             start = datetime.utcnow().timestamp()
+#             result = method(*args, **kwargs)
+#             end = datetime.utcnow().timestamp()
+#             timing(action, start, end, size)
+#         else:
+#             result = method(*args, **kwargs)
+#         return result
+#
+#     return timed
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/manage_composable.py]: Removed.

File: caom2pipe/caom2pipe/manage_composable.py [hunk 12]
```diff
@@ -983,7 +1209,8 @@ def read_obs_from_file(fqn):
 
 def read_from_file(fqn):
     """Common code to read from a text file. Mostly to make it easy to
-    mock."""
+    mock.
+    """
     if not os.path.exists(fqn):
         raise CadcException('Could not find {}'.format(fqn))
     with open(fqn, 'r') as f:
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/manage_composable.py]: :)

File: caom2pipe/caom2pipe/manage_composable.py [hunk 13]
```diff
@@ -1194,3 +1446,171 @@ def write_as_yaml(content, fqn):
...
+    :param local_fqn fully qualified name for where to file the file
+        locally.
+    """
+    try:
+        with requests.get(url, stream=True) as r:
+            r.raise_for_status()
+            with open(local_fqn, 'wb') as f:
+                for chunk in r.iter_content(chunk_size=READ_BLOCK_SIZE):
+                    f.write(chunk)
+        if not os.path.exists(local_fqn):
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/manage_composable.py]: Added checks for length and checksum headers.

These headers don't exist for VLASS, which is why there's an astropy 'verify' call in the execute_composable method that uses this method.

No retry, as that's handled elsewhere.

File: caom2pipe/caom2pipe/manage_composable.py [hunk 14]
```diff
@@ -1194,3 +1446,171 @@ def write_as_yaml(content, fqn):
...
+    try:
+        client.create(observation)
+    except Exception as e:
+        metrics.observe_failure('create', 'caom2', observation.observation_id)
+        raise CadcException(
+            'Could not create an observation record for {}. '
+            '{}'.format(observation.observation_id, e))
+    end = current()
+    metrics.observe(start, end, sizeof(observation), 'create', 'caom2',
+                    observation.observation_id)
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/manage_composable.py]: I couldn't figure out how to make a property work with a class instance. Do you know how to do that?

File: caom2pipe/caom2pipe/tests/test_run_methods.py [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+    assert os.path.exists(REJECTED_FILE)
+
+
+def cleanup_log_txt(config):
+    for fqn in [config.success_fqn, config.failure_fqn, config.retry_fqn,
+                config.rejected_fqn, config.progress_fqn]:
+        if os.path.exists(fqn):
+            os.unlink(fqn)
+    retry_dir = '{}_0'.format(TEST_DATA_DIR)
+    if os.path.exists(retry_dir):
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/tests/test_run_methods.py]: Done.

File: caom2pipe/caom2pipe/manage_composable.py [hunk 16]
```diff
@@ -1194,3 +1446,171 @@ def write_as_yaml(content, fqn):
...
+    """Deal with different time formats to get the number of
+    seconds since the epoch.
+
+    Timezone information may be present as +00, strip that for returned
+    results.
+    """
+    try:
+        index = from_time.index('+00')
+    except ValueError:
+        index = len(from_time)
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2pipe/caom2pipe/manage_composable.py]: It's ignoring the UTC timezone.

File: caom2utils/caom2utils/fits2caom2.py [hunk 17]
```diff
@@ -1647,12 +1647,16 @@ def _get_datetime(self, from_value):
                                 return datetime.strptime(
                                     from_value, '%Y-%m-%d')
                             except ValueError:
-                                self.logger.error(
-                                    'Cannot parse datetime {}'.format(
-                                        from_value))
-                                self.add_error(
-                                    'get_datetime', sys.exc_info()[1])
-                                return None
+                                try:
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2utils/caom2utils/fits2caom2.py]: For loop added.

Yes, get_seconds, and _get_datetime can probably share code, but caom2utils does not depend on caom2pipe. Should there be a refactoring to make this sharing possible?

File: caom2utils/caom2utils/tests/test_wcsvalidator.py (hunk 1 above)
Comment (SharonGoliath) [reply caom2utils/caom2utils/tests/test_wcsvalidator.py]: This is the installation:
platform linux -- Python 3.6.8, pytest-5.0.0, py-1.8.0, pluggy-0.12.0

And this is the error message:
```
E       AssertionError: assert 'unexpected TIMESYS, CTYPE' in '<ExceptionInfo InvalidWCSError tblen=2>'
E        +  where '<ExceptionInfo InvalidWCSError tblen=2>' = str(<ExceptionInfo InvalidWCSError tblen=2>)


```

**Pull request closed without merging**